packages/hebao_v1/contracts/modules/security/GuardianUtils.sol

```diff
@@ -0,0 +1,165 @@
+                walletOwnerSigned = true;
+            } else {
+                signingGuardians[numGuardians++] = securityStore.getGuardian(wallet, signers[i]);
+            }
+        }
+        // Update the signingGuardians array with the actual number of guardians that have signed
+        // (could be 1 less than the length if the owner signed as well)
+        assembly { mstore(signingGuardians, numGuardians) }
+        (
+            uint numSelfControlled,
+            uint numFriends,
+            uint numFamily,
+            uint numSocialOther
+        ) = countGuardians(signingGuardians);
+
+        if (requirement == SigRequirement.OwnerRequired) {
+            require(walletOwnerSigned, "WALLET_OWNER_SIGNATURE_REQUIRED");
+        }
+
+        if (numFriends + numFamily + numSocialOther == 0) {
+            /* Self authentication */
+            uint numSelfControlledSignersRequired = 2;
+            // If the owner is allowed to sign:
+            // - increase the total number of self controlled guardians
+            // - increase the number of required self controlled guardians
+            if (requirement != SigRequirement.OwnerNotAllowed) {
+                totalNumSelfControlled += 1;
+                numSelfControlledSignersRequired += 1;
+            }
+            // Count the wallet owner as a self controlled guardian
+            if (walletOwnerSigned) {
+                numSelfControlled += 1;
+            }
+            require(totalNumSelfControlled > 0, "SELF_AUTHENTICATION_IMPOSSIBLE");
+            if (totalNumSelfControlled >= numSelfControlledSignersRequired) {
+                require(numSelfControlled >= numSelfControlledSignersRequired, "NOT_ENOUGH_SIGNERS");
+            } else {
+                require(numSelfControlled == totalNumSelfControlled, "NOT_ENOUGH_SIGNERS");
+            }
+        } else {
+            /* Social authentication */
+            uint numGroupsActive = 0;
+            uint numGroupsSatisfied = 0;
+
+            if (totalNumFriends > 0) {
+                numGroupsActive++;
+                numGroupsSatisfied += isGroupCriteriaValid(numFriends, totalNumFriends) ? 1 : 0;
+            }
+            if (totalNumFamily > 0) {
+                numGroupsActive++;
+                numGroupsSatisfied += isGroupCriteriaValid(numFamily, totalNumFamily) ? 1 : 0;
+            }
+            if (totalNumSocialOther > 0) {
+                numGroupsActive++;
+                numGroupsSatisfied += isGroupCriteriaValid(numSocialOther, totalNumSocialOther) ? 1 : 0;
+            }
+
+            require(numGroupsActive > 0, "SOCIAL_AUTHENTICATION_IMPOSSIBLE");
+            require(numGroupsSatisfied >= (numGroupsActive*2 + 2) / 3, "NOT_ENOUGH_SIGNERS");
+        }
+    }
+
+    function isGroupCriteriaValid(
+        uint numSigners,
+        uint numTotal
+        )
+        internal
+        pure
+        returns (bool)
+    {
+        return (numSigners >= (numTotal + 1) / 2);
```

Comment:
should we use (numTotal / 2) + 1

so for an even-number of signers, we need more than 50%.